reactLab/src/InMemoryApp.js

```diff
@@ -0,0 +1,50 @@
+import {useState} from "react";
+import {generateUniqueID} from "web-vitals/dist/modules/lib/generateUniqueID";
+import App from "./App.js";
+
+function InMemoryApp(props) {
+    const [data, setData] = useState(props.initialData);
+
+    function handleItemChanged(itemID, field, newValue){
+        function edit_helper(item) {
+            if (item.id === itemID){
+                item[field] = newValue;
+                return item;
+            }
+            else {
+                return item;
+            }
+        }
+        setData(data.map(edit_helper));
+    }
+
+    function handleItemDeleted(itemID) {
+        setData(data.filter((item) => item.id !== itemID));
+    }
+
+    function handleItemCategoryDeleted(category) {
+        // delete the item based on the category they are in
+        if (category === "Completed"){
+            setData(data.filter((item) => item.completed === false));
+        } else if (category === "Uncompleted") {
+            setData(data.filter((item) => item.completed === true));
+        } else if (category === "All") {
+            setData([]);
+        }
+    }
+
+    function handleItemAdded(){
+        let newId = generateUniqueID();
```

Comment:
I'd inline newId (and possibly newItem)

Reply:
We need the newId because our itemList needs to know the newly added item (and our handleItemAdded function actually returns newId). Then the itemList will be able to tell a row if it is a newly added row. This is used to focus on the newly added item (e.g. if we have a long list, the app will autoscroll to the bottom where the new item will be focused on).